File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or k<2:
            return head
        dummy = ListNode("_")
        dummy.next = head
        p = dummy
        r = head
        count = 0
        q = r
        while q != None:
            if count == k-1:
                jp = r
                prev = r
                r = r.next
                for i in range(k-1):
                    temp = r
                    r = r.next
                    temp.next = prev
                    prev = temp
                p.next = prev
                p = jp
                jp.next = r
                q = r

                count = 1
            else:
                count += 1
            if q:
                q = q.next
        return dummy.next 

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode prev = null, curr = head, next;
        for(int i = 0; i< k; i++){
            if (curr == null){
                return head;
            }
            curr = curr.next;
        }

        curr = head;
        for(int i = 0; i< k; i++){
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        head.next = reverseKGroup(curr, k);
        return prev;
    }
}